crates/throttle/src/throttle.rs

```diff
@@ -0,0 +1,366 @@
+        &RateQuota {
+            max_burst: max_burst.min(limit - 1) as i64,
+            max_rate,
+        },
+    );
+    let quantity = quantity.unwrap_or(1) as i64;
+    let (throttled, rate_limit_result) = limiter
+        .rate_limit(key, quantity)
+        .map_err(|err| Error::Generic(format!("{err:#}")))?;
+
+    // If either time had a partial component, bump it up to the next full
+    // second because otherwise a fast-paced caller could try again too
+    // early.
+    let mut retry_after = rate_limit_result.retry_after.whole_seconds();
+    if rate_limit_result.retry_after.subsec_milliseconds() > 0 {
+        retry_after += 1
+    }
+    let mut reset_after = rate_limit_result.reset_after.whole_seconds();
+    if rate_limit_result.reset_after.subsec_milliseconds() > 0 {
+        reset_after += 1
+    }
+
+    Ok(ThrottleResult {
+        throttled,
+        limit: rate_limit_result.limit as u64,
+        remaining: rate_limit_result.remaining as u64,
+        reset_after: Duration::from_secs(reset_after.max(0) as u64),
+        retry_after: if retry_after == -1 {
+            None
+        } else {
+            Some(Duration::from_secs(retry_after.max(0) as u64))
+        },
+    })
+}
+
+async fn redis_cell_throttle(
+    conn: &RedisConnection,
+    key: &str,
+    limit: u64,
+    period: Duration,
+    max_burst: u64,
+    quantity: Option<u64>,
+) -> Result<ThrottleResult, Error> {
+    let mut cmd = Cmd::new();
+    cmd.arg("CL.THROTTLE")
+        .arg(key)
+        .arg(max_burst)
+        .arg(limit)
+        .arg(period.as_secs())
+        .arg(quantity.unwrap_or(1));
+    let result = conn.query(cmd).await?;
+    let result = <Vec<i64> as FromRedisValue>::from_redis_value(&result)?;
+
+    Ok(ThrottleResult {
+        throttled: result[0] != 0,
+        limit: result[1] as u64,
+        remaining: result[2] as u64,
+        retry_after: match result[3] {
+            n if n < 0 => None,
+            n => Some(Duration::from_secs(n as u64)),
+        },
+        reset_after: Duration::from_secs(result[4].max(0) as u64),
+    })
+}
+
+async fn redis_script_throttle(
+    conn: &RedisConnection,
+    key: &str,
+    limit: u64,
+    period: Duration,
+    max_burst: u64,
+    quantity: Option<u64>,
+) -> Result<ThrottleResult, Error> {
+    let mut script = GCRA_SCRIPT.prepare_invoke();
+    script
+        .key(key)
+        .arg(limit)
+        .arg(period.as_secs())
+        .arg(max_burst)
+        .arg(quantity.unwrap_or(1));
+
+    let result = conn
+        .invoke_script(script)
+        .await
+        .context("error invoking redis GCRA script")?;
+    let result =
+        <(u64, u64, u64, u64, String, String) as FromRedisValue>::from_redis_value(&result)?;
+
+    Ok(ThrottleResult {
+        throttled: result.0 == 1,
+        limit: max_burst + 1,
+        remaining: result.1,
+        reset_after: Duration::from_secs(result.2),
+        retry_after: match result.3 {
+            0 => None,
```

Comment:
the `redis_cell_throttle` implementation of this above handles the case where the number is negative, and also has a different representation for 0.

Which is right? TBH, I think neither is quite right!

The logic that consumes the throttles assumes that if `retry_after.is_some()`, then we were throttled, and will break out of the throttle checking loop otherwise.  It's probably best to formally encode that here for the various implementations and ensure that this is `None` if we are not throttled, and otherwise set this to the returned interval, clamping negative values to 0 so that we can't panic or return overflown/wrapped unsigned durations for this case.

I confess to not knowing whether it is possible to get a 0 second result here if we are throttled. That might be an awkward edge case that produces some busier looping in the consumer, so I'm open to considering clamping to a minimum of 1 second in the throttled case.

What do you think?

Reply:
Hmm, I noticed that `ThrottleResult::retry_after` is actually the only field from that type that gets used outside the crate. Should we just get rid of everything else?

Will dig more into the details of this tomorrow, my brain is a little too fried to go deep on this now.

Reply:
preserving the other fields seems useful to me, if for example, we wanted to have those automatically translate into the conventional set of http throttling headers in the response to an http request.

Reply:
For the script throttle, `retry_after` has `math.ceil()` applied to it, so I don't think it can return 0 when throttling happens (in that case diff must be non-zero because remaining is not 0). I think that also means it's always more than 0 in the throttled case. I've made the code a bit more similar to the `redis_cell_throttle()`.

Not sure what else to do here? Feel free to make changes to the PR if you just want to get it merged.